### DISCUSSION TITLE
Nano Els VFD issues

### DISCUSSION THREAD

**Discussion by Johannmupa** [kachurovskiy/nanoels#89] [General]
![20221120_160117](https://user-images.githubusercontent.com/118610419/202899345-92573a0c-5b0f-4d68-90f6-d71b2c4049da.jpg)

I built my Nano ELS on perfboard and used IP 67 buttons and welded a metal case that sits on top of the lathe. I use closed loop stepper powered with 60V. 
First tests were crazy. Switching on the spindle drive made the Els display to run random letters. 
Just powering the ELS and the vertical 5 button row had ghost-like inputs. Pitch was switching from 2mm to 0.2 every few seconds and multi start thread was countin up to 124 by itself. 
Today a new test with a power bank to supply the arduino. With the 60V PSU and the stepper driver still on the table not much better. Putting the PSU and driver under the table it was much better. I mounted the encoder in the 3-jaw chuck (encoder mount not ready yet). Now I was able to run the ELS and it did cut nice 2mm pitch in the air. 
When I tried the 0.2 feed it worked for a short while and switche back to 2mm by itself. 
In general Els works fine, but I have to protect it from the interferences.

**Comment by kachurovskiy** [comment]
Hi @Johannmupa, sadly this is something that happens when the PCB isn't used. PCB has the built-in ground plane that surrounds circuit lines and protects them from interference. Circuit lines are also as short as possible, minimizing their exposure.

https://user-images.githubusercontent.com/517919/202900389-72928b71-b870-4da5-bb46-c794caa63935.mp4

You could try making a "Faraday cage" for the controller by **fully** enclosing it in metal housing that is **grounded**. I didn't try this myself so can't guarantee that it will work.

In both cases you'd want to use shielded wires for stepper and encoder connections - where 4 internal wires are wrapped in metal mesh and then in plastic coating. Shielding should be connected to GND or 5V.

![image](https://user-images.githubusercontent.com/517919/202900276-d8c9fe3c-2ed9-43c9-b425-d218bbe22b80.png)

Skipping this requirement will result in extra or missing steps or spindle turns. I'm not sure if power supply wire needs to be shielded, mine isn't but I don't have a lot of EMI around my lathe.

I will also add these warnings to the main instruction doc. Thank you!

**Reply by Johannmupa** [reply]
Thank you for the quick answer. 
The case is full metal but the lid on the back not closed yet. I use shielded cable to the stepper driver and encoder. The shields connected to the board's negative. I will try tomorrow to connect the metal box to negative too and put the lid on. 
A big source of EMI might be the cable from the VFD to the 3-phase motor. I will try to find a shielded cable for it. 
My idea was to put a capacitor between +5V and ground to filter the noise. But I don't know what size to choose.

**Comment by Johannmupa** [comment]
Update
Connected -5V to metal enclosure and mounted the back lid to seal it complete.
Defenitely less issues. Just the button 2mm pitch gets activated randomly. That is the lowest voltage on AB input pin.

**Reply by kachurovskiy** [reply]
Cool. I'll see if we can do something in software to ignore noisy button clicks.

**Reply by kachurovskiy** [reply]
Maybe worth checking if all pieces of the metal box have 0 resistance to your ground wire, sometimes there's e.g. paint in the way.

**Reply by kachurovskiy** [reply]
Here's potentially a quick fix: change https://github.com/kachurovskiy/nanoels/blob/main/h2/h2.ino#L39 from `50` to `10` - this narrows the voltage area in which right-hand row of buttons triggers.

They are all on 1 analog pin, you can see their logic on https://github.com/kachurovskiy/nanoels/blob/main/h2/h2.ino#L822 - 2mm shortcut button is `B_F5`.

**Reply by Johannmupa** [reply]
Thanks a lot for your support Maxim, 
I have just discovered a video from clough42 about VFD EMI problems. His Video equipment went of when he started his VFD modified lathe. The main source oft the interference is high frequency that travels in the power line back to the wall plug. 
My guess is that it gets picked up there from the 5V adapter and travels in the USB cord to the arduino. 
Thats the reason it works much better with a power bank.
I have ordered the EMI filter that he uses quite cheap an localy so it will arive in 1or 2 days. I wil try the filter with the original code so I can see the improvement. 
I'll use the time to finaly mount my spindle encoder.

**Reply by Johannmupa** [reply]
Problem solved
After connecting this EMI filter in the power line to the VFD everything runs normal. The device is less than 10,-€ 
![0a53aa7ce900e7fe75ffc9b7ff67fa45](https://user-images.githubusercontent.com/118610419/203884244-097582df-47f6-4b67-b287-5842d0f6484b.jpg)

**Reply by kachurovskiy** [reply]
Awesome, good to know! Congratulations 👏

**Comment by Johannmupa** [comment]
Very seldom I got issues again. It only activates the 2mm pitch shortcut. I have altered the value for the voltage tolerance to 10. Then I soldered a resistor of 1 MOhm parallel to the existing 1 MOhm Resistor - this does not change the AB values significant but makes interferences more seldom. I power the NanoELS from a power bank.  Haven't had an issue for a long time. 
Today during a finishing cut the feed jumped from 0.13 to 2mm. Not a big issue. When cutting a thread the part is scrap. 
I think about deleting the 2mm pitch shortcut by giving it a very high number or find the line in the program so it takes no action. Adjusting the pitch with the up down buttons is quite fine for me.

**Reply by kachurovskiy** [reply]
Sorry to hear about that. I need to add protection from spurious noise-related button presses in the code. For now I guess you could just disable pitch shortcuts by commenting out this line: https://github.com/kachurovskiy/nanoels/blob/main/h2/h2.ino#L956

**Reply by Johannmupa** [reply]
Mary Christmas, Maxim
I guess that an VFD drive is worst case for any electronics near by. As long I don't try to weld AC at 320 Amps during threading on the late :-) 
I think tere are 2 different methods in making it more safe:
1. Disable shortcuts as long the lead screw is ON
2. A minimum time the button has to be pressed bevore it gets recognized. I guess 200mS might be ok.

**Reply by kachurovskiy** [reply]
Happy New Year, @Johannmupa :) I've added button noise protection in the latest version, please let me know if it helps.

**Comment by Johannmupa** [comment]
Dear Maxim, 
thank you for the great support. It will take some time to find out if it is failsafe now because the issues became very seldom. I have ordered an oszilloscope to find out what happens at the AB port during accellerating and braking the VFD drive. I hope to get some screenshots to load them up.

**Reply by zyeborm** [reply]
If you're still having problems a small capacitor on the input to the ADC pin of the Arduino may help reducing the input noise and it'll give the ADC something to pull charge from when it samples.  Something less than .1uf. At 0.1uf  with that 1 meg resistor the time constant is .1 seconds. So it'll take probably about half a second to really home in on the button value. 100nf is .001 second time constant so .005 seconds which is probably a lot closer to what you want. 
Something like this https://www.jaycar.com.au/100nf-50vdc-ceramic-capacitors-pack-of-2/p/RC5360 is what I'd be looking at.

**Comment by mexatroniker** [comment]
Hallo everybody,

the random key activation issue on the right is due to a circuit design error.
To eliminate this issue, as it may be caused by interference, you must:
- remove 1-kOhm resistor R7
- replace the 1-megaohm resistor R9 with a 1-kOhm resistor
- set the following trigger voltage values ​​for the analog buttons in the software code:
    #define F1_VOLTAGE 510
    #define F2_VOLTAGE 340
    #define F3_VOLTAGE 254
    #define F4_VOLTAGE 202
    #define F5_VOLTAGE 169

    #define BUTTON_EPSILON 10

Good luck!
